### ISSUE TITLE
infer.rs + HIPFIRE_GRAPH=1 panics: hipMalloc during stream capture (dense text models, gfx1100)

### ISSUE THREAD

**Issue by nickfinease** [Kaden-Schutt/hipfire#36]
## Summary

Running `examples/infer` with `HIPFIRE_GRAPH=1` on dense text models panics during prefill with `hipError 900` (`hipMalloc: operation not permitted when stream is capturing`). Reproduces on `qwen3.5-9b.mq4` and `qwen3.5-27b.mq4`. MoE A3B models don't hit this (their graph path is gated by the #19 guard).

This does **not** affect `bench_qwen35_mq4` (which has its own capture path and ignores `HIPFIRE_GRAPH`) or the daemon — so CI, speed-gate, and production inference are unaffected. Only the `infer` example harness is broken for `HIPFIRE_GRAPH=1` on dense models.

## Environment

- GPU: RX 7900 XTX (gfx1100, 24 GB)
- OS: Linux
- ROCm: 7.2
- hipfire: `e9c50c1` (post-PR #28)

## Reproducer

```bash
HIPFIRE_FP16=1 HIPFIRE_GRAPH=1 target/release/examples/infer \
  ~/.hipfire/models/qwen3.5-9b.mq4 --max-tokens 32 "Hello"
```

Panics with:

```
thread 'main' panicked at crates/engine/examples/infer.rs:213:18:
forward_scratch failed: HipError { code: 900,
message: "hipMalloc: operation not permitted when stream is capturing" }
```

Identical panic on `qwen3.5-27b.mq4`. `HIPFIRE_GRAPH=0` works fine on both.

## Cause (suspected)

`infer.rs:212` calls `qwen35::forward_scratch` inside the prefill loop. Something on the `forward_scratch` path issues a `hipMalloc` while the capture stream is live. The `bench_qwen35_mq4` harness apparently wraps the dispatch differently (pre-allocates before capture, or uses a different forward entry point) — which is why the documented v0.1.5 gfx1100 benchmark numbers (132 tok/s 9B with graph=1) were reproducible even though this path is broken.

## Impact

- Low severity — doesn't affect CI, daemon, or benchmark scripts that use `bench_qwen35_mq4`.
- Does block anyone using `examples/infer` as the "quick single-model check" suggested in `CONTRIBUTING.md` with `HIPFIRE_GRAPH=1`.
- Makes it easy to land a regression in `forward_scratch` alloc behavior silently, since only the benchmark harness exercises the captured path.

## Suggested fix direction

Either (a) hoist the allocation out of the captured region in `forward_scratch`, or (b) make `infer.rs` prefill use the same pre-allocated scratch path that `bench_qwen35_mq4` uses. (b) is probably simpler and keeps the two harnesses in sync.

Filing so the benchmark harness divergence is documented. Happy to PR once current gfx1100 benchmark submission is out the door.

**Comment by Kaden-Schutt** [comment]
Fixed in 688b4fd (master).

The crash had two layered causes, both already solved on the verify-graph path:

1. **Lazy scratch alloc under capture** (your hypothesis confirmed). The first MQ4 dispatch through `forward_scratch_layers` calls `ensure_mq_signs`, which `alloc_tensor`s the FWHT sign tables and rotation/dp4a scratch on first use. `ensure_kernel` also JIT-loads modules. Doing this inside `hipStreamBeginCapture` trips error 900. Mirrored the verify path's `verify_warmed_up` set with a scalar `Gpu.ar_forward_warmed_up` flag — first call runs direct, second captures.

2. **Sync `gpu.hip.memcpy_dtod` on null stream during capture**. After the alloc was hoisted, the next failure was error 906 ("would make the legacy stream depend on a capturing blocking stream") from the LA Q/K repeat-interleave fallback (`linear_num_key_heads == n_v_heads` branch). Routed through the existing `gpu.memcpy_dtod_auto` helper, which picks async on the active stream when set.

Verified on gfx1100 / ROCm 7.2:
- No panic on `qwen3.5-0.8b.mq4` and `qwen3.5-9b.mq4` with `HIPFIRE_FP16=1 HIPFIRE_GRAPH=1`.
- Capture instantiates 295 blobs, prefill+decode complete.
- HIPFIRE_GRAPH=0 unchanged on the coherence battery (Paris / square lambda / sheep reasoning all correct).
- `bench_qwen35_mq4` still passes with HIPFIRE_GRAPH=1.

⚠️ With the panic gone, `examples/infer` + `HIPFIRE_GRAPH=1` no longer hard-errors but the captured replay produces incoherent output on dense AR — kernels that bypass `launch_maybe_blob` (e.g. `kv_cache_write_q8_0`, several FA helpers) record stack-pointer kernargs that dangle past capture. This is the same drift class as #19 (MoE numerical divergence) and is tracked there. Not blocking your fix, since this path was unusable anyway pre-688b4fd.

Closing.

**Comment by Kaden-Schutt** [comment]
Follow-up `fbe1eb4`: added the missing `graph_launch` after `end_graph_capture`. `hipStreamCaptureModeGlobal` records but doesn't execute, so 688b4fd was leaving the first captured pos's forward unexecuted (KV write / state advance / logits all dropped). Now mirrors the verify path's explicit launch.

This doesn't fix the captured replay's coherence — kernels bypassing `launch_maybe_blob` still record dangling stack-pointer kernargs (the #19 class). But the contract for `forward_scratch` is now correct: every call advances state for its pos.
